File: backend/app/api/routes/notifications.py
import json
from datetime import UTC, datetime

from fastapi import APIRouter, Depends, HTTPException, Query, status
from sqlalchemy import and_, func, or_, select
from sqlalchemy.orm import Session

from app.core.database import get_db
from app.core.dependencies import get_current_user, require_roles
from app.models.batch import StudentBatch
from app.models.enums import NotificationType, UserRole
from app.models.fee import ReminderRule
from app.models.notification import Notification
from app.models.user import User
from app.schemas.common import PaginatedResponse
from app.schemas.notification import (
    AnnouncementCreate,
    NotificationResponse,
    ReminderRuleCreate,
    ReminderRuleResponse,
    ReminderRuleUpdate,
)
from app.services.reminder_service import generate_fee_reminders

router = APIRouter()


def _serialize_notification(notification: Notification) -> NotificationResponse:
    return NotificationResponse(
        id=notification.id,
        student_id=notification.student_id,
        batch_id=notification.batch_id,
        type=notification.type,
        message=notification.message,
        meta_json=json.loads(notification.meta_json) if notification.meta_json else None,
        created_at=notification.created_at,
        read_at=notification.read_at,
    )


def _student_batch_ids(db: Session, student_id: int) -> list[int]:
    return list(db.scalars(select(StudentBatch.batch_id).where(StudentBatch.student_id == student_id)))


def _student_can_access_notification(
    db: Session,
    student_id: int,
    notification: Notification,
) -> bool:
    if notification.student_id == student_id:
        return True
    if notification.student_id is None and notification.batch_id is None:
        return True
    if notification.student_id is None and notification.batch_id is not None:
        return notification.batch_id in _student_batch_ids(db, student_id)
    return False


@router.get("/", response_model=PaginatedResponse)
def list_notifications(
    type: NotificationType | None = Query(default=None),
    page: int = Query(default=1, ge=1),
    page_size: int = Query(default=20, ge=1, le=200),
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> PaginatedResponse:
    stmt = select(Notification).where(Notification.institute_id == current_user.institute_id)

    if type:
        stmt = stmt.where(Notification.type == type)

    if current_user.role == UserRole.STUDENT:
        if not current_user.student_id:
            raise HTTPException(status_code=status.HTTP_403_FORBIDDEN, detail="Student profile missing")
        batch_ids = list(
            db.scalars(
                select(StudentBatch.batch_id).where(StudentBatch.student_id == current_user.student_id)
            )
        )
        clauses = [
            Notification.student_id == current_user.student_id,
            and_(Notification.student_id.is_(None), Notification.batch_id.is_(None)),
        ]
        if batch_ids:
            clauses.append(and_(Notification.student_id.is_(None), Notification.batch_id.in_(batch_ids)))
        stmt = stmt.where(or_(*clauses))

    stmt = stmt.distinct()
    total = db.scalar(select(func.count()).select_from(stmt.subquery())) or 0
    notifications = list(
        db.scalars(
            stmt.order_by(Notification.created_at.desc()).offset((page - 1) * page_size).limit(page_size)
        )
    )
    return PaginatedResponse(
        total=total,
        page=page,
        page_size=page_size,
        items=[_serialize_notification(item) for item in notifications],
    )


@router.patch("/{notification_id}/read", response_model=NotificationResponse)
def mark_notification_read(
    notification_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> NotificationResponse:
    notification = db.scalar(
        select(Notification).where(
            Notification.id == notification_id,
            Notification.institute_id == current_user.institute_id,
        )
    )
    if not notification:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Notification not found")
    if current_user.role == UserRole.STUDENT:
        if not current_user.student_id:
            raise HTTPException(status_code=status.HTTP_403_FORBIDDEN, detail="Student profile missing")
        allowed = _student_can_access_notification(db, current_user.student_id, notification)
        if not allowed:
            raise HTTPException(status_code=status.HTTP_403_FORBIDDEN, detail="Not allowed")

    notification.read_at = datetime.now(UTC)
    db.commit()
    db.refresh(notification)
    return _serialize_notification(notification)


@router.post("/announcements", response_model=NotificationResponse, status_code=status.HTTP_201_CREATED)
def create_announcement(
    payload: AnnouncementCreate,
    db: Session = Depends(get_db),
    current_user: User = Depends(require_roles(UserRole.ADMIN, UserRole.TEACHER)),
) -> NotificationResponse:
    notification = Notification(
        institute_id=current_user.institute_id,
        student_id=payload.student_id,
        batch_id=payload.batch_id,
        type=NotificationType.ANNOUNCEMENT,
        message=payload.message,
    )
    db.add(notification)
    db.commit()
    db.refresh(notification)
    return _serialize_notification(notification)


@router.post("/reminder-rules", response_model=ReminderRuleResponse, status_code=status.HTTP_201_CREATED)
def create_reminder_rule(
    payload: ReminderRuleCreate,
    db: Session = Depends(get_db),
    current_user: User = Depends(require_roles(UserRole.ADMIN)),
) -> ReminderRule:
    rule = ReminderRule(
        institute_id=current_user.institute_id,
        name=payload.name,
        batch_id=payload.batch_id,
        days_before=payload.days_before,
        on_due_date=payload.on_due_date,
        every_n_days_after_due=payload.every_n_days_after_due,
        is_active=payload.is_active,
    )
    db.add(rule)
    db.commit()
    db.refresh(rule)
    return rule


@router.get("/reminder-rules", response_model=list[ReminderRuleResponse])
def list_reminder_rules(
    db: Session = Depends(get_db),
    current_user: User = Depends(require_roles(UserRole.ADMIN, UserRole.TEACHER)),
) -> list[ReminderRule]:
    return list(
        db.scalars(
            select(ReminderRule)
            .where(ReminderRule.institute_id == current_user.institute_id)
            .order_by(ReminderRule.created_at.desc())
        )
    )


@router.patch("/reminder-rules/{rule_id}", response_model=ReminderRuleResponse)
def update_reminder_rule(
    rule_id: int,
    payload: ReminderRuleUpdate,
    db: Session = Depends(get_db),
    current_user: User = Depends(require_roles(UserRole.ADMIN)),
) -> ReminderRule:
    rule = db.scalar(
        select(ReminderRule).where(ReminderRule.id == rule_id, ReminderRule.institute_id == current_user.institute_id)
    )
    if not rule:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Rule not found")
    updates = payload.model_dump(exclude_unset=True)
    for key, value in updates.items():
        setattr(rule, key, value)
    db.commit()
    db.refresh(rule)
    return rule


@router.delete("/reminder-rules/{rule_id}", status_code=status.HTTP_204_NO_CONTENT)
def delete_reminder_rule(
    rule_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(require_roles(UserRole.ADMIN)),
) -> None:
    rule = db.scalar(
        select(ReminderRule).where(ReminderRule.id == rule_id, ReminderRule.institute_id == current_user.institute_id)
    )
    if not rule:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Rule not found")
    db.delete(rule)
    db.commit()
    return None


@router.post("/run-reminders")
def run_reminders(
    db: Session = Depends(get_db),
    current_user: User = Depends(require_roles(UserRole.ADMIN, UserRole.TEACHER)),
) -> dict:
    created = generate_fee_reminders(db=db, run_date=datetime.now(UTC).date())
    db.commit()
    return {"created_notifications": created}


@router.get("/{notification_id}/whatsapp-template")
def whatsapp_template(
    notification_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> dict:
    notification = db.scalar(
        select(Notification).where(
            Notification.id == notification_id,
            Notification.institute_id == current_user.institute_id,
        )
    )
    if not notification:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Notification not found")
    if current_user.role == UserRole.STUDENT:
        if not current_user.student_id:
            raise HTTPException(status_code=status.HTTP_403_FORBIDDEN, detail="Student profile missing")
        allowed = _student_can_access_notification(db, current_user.student_id, notification)
        if not allowed:
            raise HTTPException(status_code=status.HTTP_403_FORBIDDEN, detail="Not allowed")

    meta = json.loads(notification.meta_json) if notification.meta_json else {}
    return {"template": meta.get("whatsapp_template", notification.message)}

File: backend/src/routes/notifications.js
const express = require("express");
const { connectDb } = require("../db");
const { requireAuth, requireRoles } = require("../auth");
const { parsePagination } = require("./helpers");
const { runFeeReminders } = require("../services");

const router = express.Router();

router.get("/notifications", requireAuth, async (req, res) => {
  const db = await connectDb();
  const { page, pageSize, offset } = parsePagination(req.query, 20);
  const params = [req.user.institute_id];
  let where = "WHERE institute_id = ?";
  if (req.query.type) {
    where += " AND type = ?";
    params.push(req.query.type);
  }
  if (req.user.role === "STUDENT") {
    if (!req.user.student_id) {
      return res.status(403).json({ detail: "Student profile missing" });
    }
    const links = await db.all("SELECT batch_id FROM student_batches WHERE student_id = ?", [req.user.student_id]);
    const batchIds = links.map((item) => item.batch_id);
    let clause = "(student_id = ? OR (student_id IS NULL AND batch_id IS NULL)";
    params.push(req.user.student_id);
    if (batchIds.length) {
      clause += ` OR (student_id IS NULL AND batch_id IN (${batchIds.map(() => "?").join(",")}))`;
      params.push(...batchIds);
    }
    clause += ")";
    where += ` AND ${clause}`;
  }
  const total = await db.get(`SELECT COUNT(*) as total FROM notifications ${where}`, params);
  const rows = await db.all(
    `SELECT * FROM notifications ${where}
     ORDER BY created_at DESC LIMIT ? OFFSET ?`,
    [...params, pageSize, offset]
  );
  res.json({
    total: total.total,
    page,
    page_size: pageSize,
    items: rows.map((row) => ({ ...row, meta_json: row.meta_json ? JSON.parse(row.meta_json) : null }))
  });
});

router.patch("/notifications/:notificationId/read", requireAuth, async (req, res) => {
  const db = await connectDb();
  const notificationId = Number(req.params.notificationId);
  const notification = await db.get("SELECT * FROM notifications WHERE id = ? AND institute_id = ?", [notificationId, req.user.institute_id]);
  if (!notification) {
    return res.status(404).json({ detail: "Notification not found" });
  }
  if (req.user.role === "STUDENT") {
    if (!req.user.student_id) {
      return res.status(403).json({ detail: "Student profile missing" });
    }
    const links = await db.all("SELECT batch_id FROM student_batches WHERE student_id = ?", [req.user.student_id]);
    const batchSet = new Set(links.map((item) => item.batch_id));
    const allowed =
      notification.student_id === req.user.student_id ||
      (notification.student_id == null && notification.batch_id == null) ||
      (notification.student_id == null && notification.batch_id != null && batchSet.has(notification.batch_id));
    if (!allowed) {
      return res.status(403).json({ detail: "Not allowed" });
    }
  }
  await db.run("UPDATE notifications SET read_at = CURRENT_TIMESTAMP WHERE id = ?", [notificationId]);
  const updated = await db.get("SELECT * FROM notifications WHERE id = ?", [notificationId]);
  res.json({ ...updated, meta_json: updated.meta_json ? JSON.parse(updated.meta_json) : null });
});

router.post("/notifications/announcements", requireRoles("ADMIN", "TEACHER"), async (req, res) => {
  const db = await connectDb();
  const created = await db.run(
    `INSERT INTO notifications (institute_id, student_id, batch_id, type, message)
     VALUES (?, ?, ?, 'ANNOUNCEMENT', ?)`,
    [req.user.institute_id, req.body.student_id || null, req.body.batch_id || null, req.body.message]
  );
  const row = await db.get("SELECT * FROM notifications WHERE id = ?", [created.lastID]);
  res.status(201).json({ ...row, meta_json: null });
});

router.post("/notifications/reminder-rules", requireRoles("ADMIN"), async (req, res) => {
  const db = await connectDb();
  const created = await db.run(
    `INSERT INTO reminder_rules (institute_id, batch_id, name, days_before, on_due_date, every_n_days_after_due, is_active)
     VALUES (?, ?, ?, ?, ?, ?, ?)`,
    [
      req.user.institute_id,
      req.body.batch_id || null,
      req.body.name,
      req.body.days_before,
      req.body.on_due_date ? 1 : 0,
      req.body.every_n_days_after_due,
      req.body.is_active == null ? 1 : req.body.is_active ? 1 : 0
    ]
  );
  const rule = await db.get("SELECT * FROM reminder_rules WHERE id = ?", [created.lastID]);
  rule.on_due_date = Boolean(rule.on_due_date);
  rule.is_active = Boolean(rule.is_active);
  res.status(201).json(rule);
});

router.get("/notifications/reminder-rules", requireRoles("ADMIN", "TEACHER"), async (req, res) => {
  const db = await connectDb();
  const rows = await db.all("SELECT * FROM reminder_rules WHERE institute_id = ? ORDER BY created_at DESC", [req.user.institute_id]);
  res.json(rows.map((row) => ({ ...row, on_due_date: Boolean(row.on_due_date), is_active: Boolean(row.is_active) })));
});

router.patch("/notifications/reminder-rules/:ruleId", requireRoles("ADMIN"), async (req, res) => {
  const db = await connectDb();
  const ruleId = Number(req.params.ruleId);
  const row = await db.get("SELECT id FROM reminder_rules WHERE id = ? AND institute_id = ?", [ruleId, req.user.institute_id]);
  if (!row) {
    return res.status(404).json({ detail: "Rule not found" });
  }
  const updates = { ...req.body };
  if (updates.on_due_date !== undefined) {
    updates.on_due_date = updates.on_due_date ? 1 : 0;
  }
  if (updates.is_active !== undefined) {
    updates.is_active = updates.is_active ? 1 : 0;
  }
  Object.keys(updates).forEach((key) => updates[key] === undefined && delete updates[key]);
  if (Object.keys(updates).length) {
    const setSql = Object.keys(updates).map((key) => `${key} = ?`).join(", ");
    await db.run(`UPDATE reminder_rules SET ${setSql}, updated_at = CURRENT_TIMESTAMP WHERE id = ?`, [
      ...Object.values(updates),
      ruleId
    ]);
  }
  const updated = await db.get("SELECT * FROM reminder_rules WHERE id = ?", [ruleId]);
  updated.on_due_date = Boolean(updated.on_due_date);
  updated.is_active = Boolean(updated.is_active);
  res.json(updated);
});

router.delete("/notifications/reminder-rules/:ruleId", requireRoles("ADMIN"), async (req, res) => {
  const db = await connectDb();
  const ruleId = Number(req.params.ruleId);
  const row = await db.get("SELECT id FROM reminder_rules WHERE id = ? AND institute_id = ?", [ruleId, req.user.institute_id]);
  if (!row) {
    return res.status(404).json({ detail: "Rule not found" });
  }
  await db.run("DELETE FROM reminder_rules WHERE id = ?", [ruleId]);
  res.status(204).send();
});

router.post("/notifications/run-reminders", requireRoles("ADMIN", "TEACHER"), async (_req, res) => {
  const created = await runFeeReminders(new Date().toISOString().slice(0, 10));
  res.json({ created_notifications: created });
});

router.get("/notifications/:notificationId/whatsapp-template", requireAuth, async (req, res) => {
  const db = await connectDb();
  const row = await db.get("SELECT * FROM notifications WHERE id = ? AND institute_id = ?", [Number(req.params.notificationId), req.user.institute_id]);
  if (!row) {
    return res.status(404).json({ detail: "Notification not found" });
  }
  if (req.user.role === "STUDENT") {
    if (!req.user.student_id) {
      return res.status(403).json({ detail: "Student profile missing" });
    }
    const links = await db.all("SELECT batch_id FROM student_batches WHERE student_id = ?", [req.user.student_id]);
    const batchSet = new Set(links.map((item) => item.batch_id));
    const allowed =
      row.student_id === req.user.student_id ||
      (row.student_id == null && row.batch_id == null) ||
      (row.student_id == null && row.batch_id != null && batchSet.has(row.batch_id));
    if (!allowed) {
      return res.status(403).json({ detail: "Not allowed" });
    }
  }
  const meta = row.meta_json ? JSON.parse(row.meta_json) : {};
  res.json({ template: meta.whatsapp_template || row.message });
});

module.exports = router;
